docs/proposals/provider-agnostic-dns-health-checks.md

```diff
@@ -0,0 +1,89 @@
+# Provider agnostic DNS Health checks
+
+## Introduction
+
+The MGC has the ability to extend the DNS configuration of the gateway with the DNSPolicy resource. This resource allows 
+users to configure health checks. As a result of configuring health checks, the controller creates the health checks in 
+Route53, attaching them to the related DNS records. This has the benefit of automatically disabling an endpoint if it 
+becomes unhealthy, and enabling it again when it becomes healthy again.
+
+This feature has a few shortfalls:
+1. It’s tightly coupled with Route53. If other DNS providers are supported they must either provide a similar feature, 
+or health checks will not be supported
+2. Lacks the ability to reach endpoints in private networks
+
+This document describes a proposal to extend the current health check implementation to overcome these shortfalls.
+
+### Goals
+
+* Ability to configure health checks in the DNSPolicy associated to a Gateway
+* DNS records are disabled when the associated health check fails
+* Current status of the defined health checks is visible to the end user
+
+### Nongoals
+* Ability for the health checks to reach endpoints in private networks
+* Transparently keep support for other health check providers like Route53
+
+## Proposal
+
+#### `DNSPolicy` resource
+
+The presence of the `healthCheck` means that for every DNS endpoint (that is either an A record, or a CNAME to an external host), 
+a health check is created based on the health check configuration in the DNSPolicy.
+
+A `failureThreshold` field will be added to the health spec, allowing users to configure a number of consecutive health 
+check failures that must be observed before the endpoint is considered unhealthy.
+
+#### `DNSHealthCheckProbe` resource
+
+The DNSHealthCheckProbe resource configures a health probe in the controller to perform the health checks against an 
+endpoint.
+
+```yaml
+apiVersion: kuadrant.io/v1alpha1
+kind: DNSHealthCheckProbe
+metadata:
+  name: example-probe
+spec:
+  port: "..."
+  host: “...”
+  address: "..."
+  path: "..."
+  protocol: "..."
+  interval: "..."
+status:
+  healthy: true
+  consecutiveFailures: 0
+  reason: ""
+  lastCheck: "..."
+```
+
+#### Spec Fields Definition
+- **Port** The port to use
+- **Address** The address to connect to (e.g. IP address or hostname of a clusters loadbalancer)
+- **Host** The host to request in the Host header
+- **Path** The path to request
+- **Protocol** The protocol to use for this request
+- **Interval** How frequently this check would ideally be executed.
+
+The reconciliation of this resource results in the configuration of a health probe, which targets the endpoint and 
+updates the status. The status is propagated to the providerSpecific status of the equivalent endpoint in the DNSRecord
+
+### Changes to current controllers
```

Comment:
do we not want to outline or call out that there will be a new health check agent? Seems like the fundamental proposal here is to introduce a new health check agent component?

Reply:
I think we should also at a high level explain how the agent will perform the health checks and where the agent will live

Reply:
There is no "seperate" agent now, right? Or I've significantly misunderstood our most recent discussion. The code is running inside the existing controller, rate limited by a queue and running on a separate goroutine.

Reply:
ok I guess a separate service is not needed at this point. But I can see the potential need for it in the future if we want to install into the spoke or scale independently. Really this means adding a separate main and coming up with a mechanism for deploying it. So as they can share the same code base, initially having the health checks start as part of the main controller seems fine for the initial release. Lets capture this in the proposal though